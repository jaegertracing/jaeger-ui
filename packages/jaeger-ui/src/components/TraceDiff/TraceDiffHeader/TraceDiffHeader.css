/*
Copyright (c) 2018 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.TraceDiffHeader {
  background: #fff;
  border-bottom: 1px solid #e8e8e8;
  box-shadow: 0 1px 4px rgba(0, 0, 0, 0.1);
  display: flex;
  flex: 0;
  position: relative;
  z-index: 1;
  align-items: stretch;
}

.TraceDiffHeader--labelItem {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 50px;
  background: #f5f5f5;
  border-right: 1px solid #e8e8e8;
}

.TraceDiffHeader--label {
  margin: 0;
  font-size: 24px;
  font-weight: 600;
  color: #333;
}

.TraceDiffHeader--traceSection {
  display: flex;
  flex: 1;
  min-width: 0;
  cursor: pointer;
  transition: background-color 0.2s ease;
}

.TraceDiffHeader--traceSection:hover {
  background-color: #f9f9f9;
}

.TraceDiffHeader--divider {
  display: flex;
  align-items: center;
  justify-content: center;
  padding: 0 15px;
  background: #f5f5f5;
  border-right: 1px solid #e8e8e8;
  border-left: 1px solid #e8e8e8;
}

.TraceDiffHeader--vsContainer {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 36px;
  height: 36px;
  border-radius: 50%;
  background: #199;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.15);
  transition:
    transform 0.2s ease,
    box-shadow 0.2s ease;
}

.TraceDiffHeader--vsContainer:hover {
  transform: scale(1.05);
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.2);
}

.TraceDiffHeader--vsLabel {
  font-size: 14px;
  font-weight: 700;
  color: white;
  letter-spacing: 1px;
  text-shadow: 0 1px 1px rgba(0, 0, 0, 0.2);
}

.TraceDiffHeader--popover .ant-popover-title {
  padding: 10px;
  border-bottom: 1px solid #e8e8e8;
}

.TraceDiffHeader--popover .ant-popover-inner-content {
  padding: 0;
}

.TraceDiffHeader--popover .ant-popover-inner {
  border-radius: 4px;
  box-shadow:
    0 3px 6px -4px rgba(0, 0, 0, 0.12),
    0 6px 16px 0 rgba(0, 0, 0, 0.08),
    0 9px 28px 8px rgba(0, 0, 0, 0.05);
}
