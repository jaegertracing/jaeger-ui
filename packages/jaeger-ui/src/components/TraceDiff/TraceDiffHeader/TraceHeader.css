/*
Copyright (c) 2017 Uber Technologies, Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.TraecDiffHeader--traceHeader {
  align-items: stretch;
  background: #fafafa;
  border-right: 1px solid #d8d8d8;
  display: flex;
  flex-direction: column;
  flex: 1;
}

.TraecDiffHeader--traceTitle {
  align-items: center;
  background: #fff;
  border-bottom: 1px solid #eee;
  display: flex;
  flex: 1;
  font-size: 1.8em;
  margin: 0;
  padding: 0.25rem 2.5rem 0.25rem 1.25rem;
  position: relative;
}

.TraecDiffHeader--traceTitle.is-error {
  color: #c00;
}

.TraecDiffHeader--traceTitleChevron {
  color: var(--tx-color-title);
  font-size: 0.75em;
  margin-right: 0.75em;
  position: absolute;
  right: 0;
  top: calc(50% - 0.5em);
}

.TraecDiffHeader--traceAttributes {
  list-style: none;
  margin: 0;
  padding: 0.25rem 1.25rem;
}

.TraecDiffHeader--traceAttr {
  display: inline-block;
  margin-right: 1rem;
}
