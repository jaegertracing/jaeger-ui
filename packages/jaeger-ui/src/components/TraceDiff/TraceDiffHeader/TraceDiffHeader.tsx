// Copyright (c) 2017 Uber Technologies, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import * as React from 'react';
import { Popover } from 'antd';

import CohortTable from './CohortTable';
import TraceHeader from './TraceHeader';
import TraceIdInput from './TraceIdInput';

import { FetchedTrace, TNil } from '../../../types';

import './TraceDiffHeader.css';

type Props = {
  a: FetchedTrace | TNil;
  b: FetchedTrace | TNil;
  cohort: FetchedTrace[];
  diffSetA: (traceId: string) => void;
  diffSetB: (traceId: string) => void;
};

type State = {
  tableVisible: ('a' | 'b') | null;
};

export default class TraceDiffHeader extends React.PureComponent<Props, State> {
  _toggleTableA: (showTable: boolean) => void;
  _toggleTableB: (showTable: boolean) => void;
  _diffSetA: (traceID: string) => void;
  _diffSetB: (traceID: string) => void;

  state = {
    tableVisible: null,
  };

  constructor(props: Props) {
    super(props);
    this._toggleTableA = this._toggleTable.bind(this, 'a');
    this._toggleTableB = this._toggleTable.bind(this, 'b');
    this._diffSetA = this._diffSetTrace.bind(this, 'a');
    this._diffSetB = this._diffSetTrace.bind(this, 'b');
  }

  _toggleTable(which: 'a' | 'b', visible: boolean) {
    const tableVisible = visible ? which : null;
    this.setState({ tableVisible });
  }

  _diffSetTrace(which: 'a' | 'b', id: string) {
    if (which === 'a') {
      this.props.diffSetA(id);
    } else {
      this.props.diffSetB(id);
    }
    this.setState({ tableVisible: null });
  }

  render() {
    const { a, b, cohort } = this.props;
    const { tableVisible } = this.state;
    const {
      data: aData = undefined,
      id: aId = undefined,
      state: aState = undefined,
      error: aError = undefined,
    } = a || {};
    const {
      data: bData = undefined,
      id: bId = undefined,
      state: bState = undefined,
      error: bError = undefined,
    } = b || {};
    const selection: Record<string, { label: 'A' | 'B' }> = {};
    if (aId) selection[aId] = { label: 'A' };
    if (bId) selection[bId] = { label: 'B' };
    const cohortTableA = (
      <CohortTable cohort={cohort} current={aId} selectTrace={this._diffSetA} selection={selection} />
    );
    const cohortTableB = (
      <CohortTable cohort={cohort} current={bId} selectTrace={this._diffSetB} selection={selection} />
    );
    return (
      <header className="TraecDiffHeader">
        <div className="TraecDiffHeader--labelItem-darkened">
          <h1 className="ub-m0">A</h1>
        </div>
        <Popover
          overlayClassName="TraceDiffHeader--popover"
          trigger="click"
          placement="bottomLeft"
          title={<TraceIdInput selectTrace={this._diffSetA} />}
          content={cohortTableA}
          open={tableVisible === 'a'}
          onOpenChange={this._toggleTableA}
        >
          <div className="ub-flex u-flex-1">
            <TraceHeader
              duration={aData && aData.duration}
              error={aError}
              startTime={aData && aData.startTime}
              state={aState}
              totalSpans={aData && aData.spans && aData.spans.length}
              traceID={aId}
              traceName={aData && aData.traceName}
            />
          </div>
        </Popover>
        <div className="TraecDiffHeader--labelItem">
          <h2 className="u-tx-muted ub-mx3 ub-my0">VS</h2>
        </div>
        <div className="TraecDiffHeader--labelItem-darkened">
          <h1 className="ub-m0">B</h1>
        </div>
        <Popover
          overlayClassName="TraceDiffHeader--popover"
          trigger="click"
          placement="bottomLeft"
          title={<TraceIdInput selectTrace={this._diffSetB} />}
          content={cohortTableB}
          open={tableVisible === 'b'}
          onOpenChange={this._toggleTableB}
        >
          <div className="ub-flex u-flex-1">
            <TraceHeader
              duration={bData && bData.duration}
              error={bError}
              startTime={bData && bData.startTime}
              state={bState}
              totalSpans={bData && bData.spans && bData.spans.length}
              traceID={bId}
              traceName={bData && bData.traceName}
            />
          </div>
        </Popover>
      </header>
    );
  }
}
