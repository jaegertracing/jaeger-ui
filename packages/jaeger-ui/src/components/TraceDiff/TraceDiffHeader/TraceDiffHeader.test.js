// Copyright (c) 2019 The Jaeger Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import React from 'react';
import { shallow } from 'enzyme';
import { Popover } from 'antd';

import TraceDiffHeader from './TraceDiffHeader';
import { fetchedState } from '../../../constants';

describe('TraceDiffHeader', () => {
  const cohort = [
    {
      data: {
        duration: 0,
        // purposefully missing spans
        startTime: 0,
        traceName: 'cohort-trace-name-0',
      },
      error: 'error 0',
      id: 'cohort-id-0',
      state: fetchedState.ERROR,
    },
    {
      data: {
        duration: 100,
        spans: [
          {
            spanID: 'trace-1-span-0',
          },
        ],
        startTime: 100,
        traceName: 'cohort-trace-name-1',
      },
      error: 'error 1',
      id: 'cohort-id-1',
      state: fetchedState.DONE,
    },
    {
      data: {
        duration: 200,
        spans: [
          {
            spanID: 'trace-2-span-1',
          },
          {
            spanID: 'trace-2-span-2',
          },
        ],
        startTime: 200,
        traceName: 'cohort-trace-name-2',
      },
      error: 'error 2',
      id: 'cohort-id-2',
      state: fetchedState.DONE,
    },
    {
      data: {
        duration: 300,
        spans: [
          {
            spanID: 'trace-3-span-1',
          },
          {
            spanID: 'trace-3-span-2',
          },
          {
            spanID: 'trace-3-span-3',
          },
        ],
        startTime: 300,
        traceName: 'cohort-trace-name-3',
      },
      error: 'error 3',
      id: 'cohort-id-3',
      state: fetchedState.DONE,
    },
  ];
  const diffSetA = jest.fn();
  const diffSetB = jest.fn();
  const props = {
    a: cohort[1],
    b: cohort[2],
    cohort,
    diffSetA,
    diffSetB,
  };

  let wrapper;

  function getPopoverProp(popoverIndex, propName) {
    return wrapper
      .find(Popover)
      .at(popoverIndex)
      .prop(propName);
  }

  beforeEach(() => {
    diffSetA.mockReset();
    diffSetB.mockReset();
    wrapper = shallow(<TraceDiffHeader {...props} />);
  });

  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('handles trace without spans', () => {
    wrapper.setProps({ a: cohort[0] });
  });

  it('handles absent a', () => {
    wrapper.setProps({ a: null });
    expect(wrapper).toMatchSnapshot();
  });

  it('handles absent b', () => {
    wrapper.setProps({ b: null });
    expect(wrapper).toMatchSnapshot();
  });

  it('handles absent a & b', () => {
    wrapper.setProps({ a: null, b: null });
    expect(wrapper).toMatchSnapshot();
  });

  it('manages visibility correctly', () => {
    expect(wrapper.state().tableVisible).toBe(null);
    const popovers = wrapper.find(Popover);
    expect(popovers.length).toBe(2);
    popovers.forEach(popover => expect(popover.prop('visible')).toBe(false));

    getPopoverProp(0, 'onVisibleChange')(true);
    expect(getPopoverProp(0, 'visible')).toBe(true);
    expect(getPopoverProp(1, 'visible')).toBe(false);

    getPopoverProp(1, 'onVisibleChange')(true);
    expect(getPopoverProp(0, 'visible')).toBe(false);
    expect(getPopoverProp(1, 'visible')).toBe(true);

    // repeat onVisibleChange call to test that visibility remains correct
    getPopoverProp(1, 'onVisibleChange')(true);
    expect(getPopoverProp(0, 'visible')).toBe(false);
    expect(getPopoverProp(1, 'visible')).toBe(true);

    getPopoverProp(1, 'onVisibleChange')(false);
    wrapper.find(Popover).forEach(popover => expect(popover.prop('visible')).toBe(false));
  });

  describe('bound functions to set a & b and passes them to Popover JSX props correctly', () => {
    const shouldCall = {
      a: diffSetA,
      b: diffSetB,
    };
    const shouldNotCall = {
      a: diffSetB,
      b: diffSetA,
    };

    ['a', 'b'].forEach(aOrB => {
      ['title', 'content'].forEach(popoverSection => {
        it(`sets trace ${aOrB} from popover ${popoverSection}`, () => {
          const selectTraceArgument = `aOrB: ${aOrB}, popoverSection: ${popoverSection}`;
          wrapper.setState({ tableVisible: aOrB });
          wrapper
            .find(Popover)
            .at(Number(aOrB === 'b'))
            .prop(popoverSection)
            .props.selectTrace(selectTraceArgument);

          expect(shouldCall[aOrB]).toHaveBeenCalledWith(selectTraceArgument);
          expect(shouldNotCall[aOrB]).not.toHaveBeenCalled();
          expect(wrapper.state().tableVisible).toBe(null);
        });
      });
    });
  });
});
