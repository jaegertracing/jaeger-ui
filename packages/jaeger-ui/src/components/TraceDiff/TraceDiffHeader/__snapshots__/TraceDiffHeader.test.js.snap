// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TraceDiffHeader handles absent a & b 1`] = `
<header
  className="TraecDiffHeader"
>
  <div
    className="TraecDiffHeader--labelItem-darkened"
  >
    <h1
      className="ub-m0"
    >
      A
    </h1>
  </div>
  <Popover
    content={
      <CohortTable
        cohort={
          Array [
            Object {
              "data": Object {
                "duration": 0,
                "startTime": 0,
                "traceName": "cohort-trace-name-0",
              },
              "error": "error 0",
              "id": "cohort-id-0",
              "state": "FETCH_ERROR",
            },
            Object {
              "data": Object {
                "duration": 100,
                "spans": Array [
                  Object {
                    "spanID": "trace-1-span-0",
                  },
                ],
                "startTime": 100,
                "traceName": "cohort-trace-name-1",
              },
              "error": "error 1",
              "id": "cohort-id-1",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 200,
                "spans": Array [
                  Object {
                    "spanID": "trace-2-span-1",
                  },
                  Object {
                    "spanID": "trace-2-span-2",
                  },
                ],
                "startTime": 200,
                "traceName": "cohort-trace-name-2",
              },
              "error": "error 2",
              "id": "cohort-id-2",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 300,
                "spans": Array [
                  Object {
                    "spanID": "trace-3-span-1",
                  },
                  Object {
                    "spanID": "trace-3-span-2",
                  },
                  Object {
                    "spanID": "trace-3-span-3",
                  },
                ],
                "startTime": 300,
                "traceName": "cohort-trace-name-3",
              },
              "error": "error 3",
              "id": "cohort-id-3",
              "state": "FETCH_DONE",
            },
          ]
        }
        selectTrace={[Function]}
        selection={Object {}}
      />
    }
    onVisibleChange={[Function]}
    overlayClassName="TraceDiffHeader--popover"
    placement="bottomLeft"
    title={
      <TraceIdInput
        selectTrace={[Function]}
      />
    }
    trigger="click"
    visible={false}
  >
    <div
      className="ub-flex u-flex-1"
    >
      <TraceHeader />
    </div>
  </Popover>
  <div
    className="TraecDiffHeader--labelItem"
  >
    <h2
      className="u-tx-muted ub-mx3 ub-my0"
    >
      VS
    </h2>
  </div>
  <div
    className="TraecDiffHeader--labelItem-darkened"
  >
    <h1
      className="ub-m0"
    >
      B
    </h1>
  </div>
  <Popover
    content={
      <CohortTable
        cohort={
          Array [
            Object {
              "data": Object {
                "duration": 0,
                "startTime": 0,
                "traceName": "cohort-trace-name-0",
              },
              "error": "error 0",
              "id": "cohort-id-0",
              "state": "FETCH_ERROR",
            },
            Object {
              "data": Object {
                "duration": 100,
                "spans": Array [
                  Object {
                    "spanID": "trace-1-span-0",
                  },
                ],
                "startTime": 100,
                "traceName": "cohort-trace-name-1",
              },
              "error": "error 1",
              "id": "cohort-id-1",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 200,
                "spans": Array [
                  Object {
                    "spanID": "trace-2-span-1",
                  },
                  Object {
                    "spanID": "trace-2-span-2",
                  },
                ],
                "startTime": 200,
                "traceName": "cohort-trace-name-2",
              },
              "error": "error 2",
              "id": "cohort-id-2",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 300,
                "spans": Array [
                  Object {
                    "spanID": "trace-3-span-1",
                  },
                  Object {
                    "spanID": "trace-3-span-2",
                  },
                  Object {
                    "spanID": "trace-3-span-3",
                  },
                ],
                "startTime": 300,
                "traceName": "cohort-trace-name-3",
              },
              "error": "error 3",
              "id": "cohort-id-3",
              "state": "FETCH_DONE",
            },
          ]
        }
        selectTrace={[Function]}
        selection={Object {}}
      />
    }
    onVisibleChange={[Function]}
    overlayClassName="TraceDiffHeader--popover"
    placement="bottomLeft"
    title={
      <TraceIdInput
        selectTrace={[Function]}
      />
    }
    trigger="click"
    visible={false}
  >
    <div
      className="ub-flex u-flex-1"
    >
      <TraceHeader />
    </div>
  </Popover>
</header>
`;

exports[`TraceDiffHeader handles absent a 1`] = `
<header
  className="TraecDiffHeader"
>
  <div
    className="TraecDiffHeader--labelItem-darkened"
  >
    <h1
      className="ub-m0"
    >
      A
    </h1>
  </div>
  <Popover
    content={
      <CohortTable
        cohort={
          Array [
            Object {
              "data": Object {
                "duration": 0,
                "startTime": 0,
                "traceName": "cohort-trace-name-0",
              },
              "error": "error 0",
              "id": "cohort-id-0",
              "state": "FETCH_ERROR",
            },
            Object {
              "data": Object {
                "duration": 100,
                "spans": Array [
                  Object {
                    "spanID": "trace-1-span-0",
                  },
                ],
                "startTime": 100,
                "traceName": "cohort-trace-name-1",
              },
              "error": "error 1",
              "id": "cohort-id-1",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 200,
                "spans": Array [
                  Object {
                    "spanID": "trace-2-span-1",
                  },
                  Object {
                    "spanID": "trace-2-span-2",
                  },
                ],
                "startTime": 200,
                "traceName": "cohort-trace-name-2",
              },
              "error": "error 2",
              "id": "cohort-id-2",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 300,
                "spans": Array [
                  Object {
                    "spanID": "trace-3-span-1",
                  },
                  Object {
                    "spanID": "trace-3-span-2",
                  },
                  Object {
                    "spanID": "trace-3-span-3",
                  },
                ],
                "startTime": 300,
                "traceName": "cohort-trace-name-3",
              },
              "error": "error 3",
              "id": "cohort-id-3",
              "state": "FETCH_DONE",
            },
          ]
        }
        selectTrace={[Function]}
        selection={
          Object {
            "cohort-id-2": Object {
              "label": "B",
            },
          }
        }
      />
    }
    onVisibleChange={[Function]}
    overlayClassName="TraceDiffHeader--popover"
    placement="bottomLeft"
    title={
      <TraceIdInput
        selectTrace={[Function]}
      />
    }
    trigger="click"
    visible={false}
  >
    <div
      className="ub-flex u-flex-1"
    >
      <TraceHeader />
    </div>
  </Popover>
  <div
    className="TraecDiffHeader--labelItem"
  >
    <h2
      className="u-tx-muted ub-mx3 ub-my0"
    >
      VS
    </h2>
  </div>
  <div
    className="TraecDiffHeader--labelItem-darkened"
  >
    <h1
      className="ub-m0"
    >
      B
    </h1>
  </div>
  <Popover
    content={
      <CohortTable
        cohort={
          Array [
            Object {
              "data": Object {
                "duration": 0,
                "startTime": 0,
                "traceName": "cohort-trace-name-0",
              },
              "error": "error 0",
              "id": "cohort-id-0",
              "state": "FETCH_ERROR",
            },
            Object {
              "data": Object {
                "duration": 100,
                "spans": Array [
                  Object {
                    "spanID": "trace-1-span-0",
                  },
                ],
                "startTime": 100,
                "traceName": "cohort-trace-name-1",
              },
              "error": "error 1",
              "id": "cohort-id-1",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 200,
                "spans": Array [
                  Object {
                    "spanID": "trace-2-span-1",
                  },
                  Object {
                    "spanID": "trace-2-span-2",
                  },
                ],
                "startTime": 200,
                "traceName": "cohort-trace-name-2",
              },
              "error": "error 2",
              "id": "cohort-id-2",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 300,
                "spans": Array [
                  Object {
                    "spanID": "trace-3-span-1",
                  },
                  Object {
                    "spanID": "trace-3-span-2",
                  },
                  Object {
                    "spanID": "trace-3-span-3",
                  },
                ],
                "startTime": 300,
                "traceName": "cohort-trace-name-3",
              },
              "error": "error 3",
              "id": "cohort-id-3",
              "state": "FETCH_DONE",
            },
          ]
        }
        current="cohort-id-2"
        selectTrace={[Function]}
        selection={
          Object {
            "cohort-id-2": Object {
              "label": "B",
            },
          }
        }
      />
    }
    onVisibleChange={[Function]}
    overlayClassName="TraceDiffHeader--popover"
    placement="bottomLeft"
    title={
      <TraceIdInput
        selectTrace={[Function]}
      />
    }
    trigger="click"
    visible={false}
  >
    <div
      className="ub-flex u-flex-1"
    >
      <TraceHeader
        duration={200}
        error="error 2"
        startTime={200}
        state="FETCH_DONE"
        totalSpans={2}
        traceID="cohort-id-2"
        traceName="cohort-trace-name-2"
      />
    </div>
  </Popover>
</header>
`;

exports[`TraceDiffHeader handles absent b 1`] = `
<header
  className="TraecDiffHeader"
>
  <div
    className="TraecDiffHeader--labelItem-darkened"
  >
    <h1
      className="ub-m0"
    >
      A
    </h1>
  </div>
  <Popover
    content={
      <CohortTable
        cohort={
          Array [
            Object {
              "data": Object {
                "duration": 0,
                "startTime": 0,
                "traceName": "cohort-trace-name-0",
              },
              "error": "error 0",
              "id": "cohort-id-0",
              "state": "FETCH_ERROR",
            },
            Object {
              "data": Object {
                "duration": 100,
                "spans": Array [
                  Object {
                    "spanID": "trace-1-span-0",
                  },
                ],
                "startTime": 100,
                "traceName": "cohort-trace-name-1",
              },
              "error": "error 1",
              "id": "cohort-id-1",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 200,
                "spans": Array [
                  Object {
                    "spanID": "trace-2-span-1",
                  },
                  Object {
                    "spanID": "trace-2-span-2",
                  },
                ],
                "startTime": 200,
                "traceName": "cohort-trace-name-2",
              },
              "error": "error 2",
              "id": "cohort-id-2",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 300,
                "spans": Array [
                  Object {
                    "spanID": "trace-3-span-1",
                  },
                  Object {
                    "spanID": "trace-3-span-2",
                  },
                  Object {
                    "spanID": "trace-3-span-3",
                  },
                ],
                "startTime": 300,
                "traceName": "cohort-trace-name-3",
              },
              "error": "error 3",
              "id": "cohort-id-3",
              "state": "FETCH_DONE",
            },
          ]
        }
        current="cohort-id-1"
        selectTrace={[Function]}
        selection={
          Object {
            "cohort-id-1": Object {
              "label": "A",
            },
          }
        }
      />
    }
    onVisibleChange={[Function]}
    overlayClassName="TraceDiffHeader--popover"
    placement="bottomLeft"
    title={
      <TraceIdInput
        selectTrace={[Function]}
      />
    }
    trigger="click"
    visible={false}
  >
    <div
      className="ub-flex u-flex-1"
    >
      <TraceHeader
        duration={100}
        error="error 1"
        startTime={100}
        state="FETCH_DONE"
        totalSpans={1}
        traceID="cohort-id-1"
        traceName="cohort-trace-name-1"
      />
    </div>
  </Popover>
  <div
    className="TraecDiffHeader--labelItem"
  >
    <h2
      className="u-tx-muted ub-mx3 ub-my0"
    >
      VS
    </h2>
  </div>
  <div
    className="TraecDiffHeader--labelItem-darkened"
  >
    <h1
      className="ub-m0"
    >
      B
    </h1>
  </div>
  <Popover
    content={
      <CohortTable
        cohort={
          Array [
            Object {
              "data": Object {
                "duration": 0,
                "startTime": 0,
                "traceName": "cohort-trace-name-0",
              },
              "error": "error 0",
              "id": "cohort-id-0",
              "state": "FETCH_ERROR",
            },
            Object {
              "data": Object {
                "duration": 100,
                "spans": Array [
                  Object {
                    "spanID": "trace-1-span-0",
                  },
                ],
                "startTime": 100,
                "traceName": "cohort-trace-name-1",
              },
              "error": "error 1",
              "id": "cohort-id-1",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 200,
                "spans": Array [
                  Object {
                    "spanID": "trace-2-span-1",
                  },
                  Object {
                    "spanID": "trace-2-span-2",
                  },
                ],
                "startTime": 200,
                "traceName": "cohort-trace-name-2",
              },
              "error": "error 2",
              "id": "cohort-id-2",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 300,
                "spans": Array [
                  Object {
                    "spanID": "trace-3-span-1",
                  },
                  Object {
                    "spanID": "trace-3-span-2",
                  },
                  Object {
                    "spanID": "trace-3-span-3",
                  },
                ],
                "startTime": 300,
                "traceName": "cohort-trace-name-3",
              },
              "error": "error 3",
              "id": "cohort-id-3",
              "state": "FETCH_DONE",
            },
          ]
        }
        selectTrace={[Function]}
        selection={
          Object {
            "cohort-id-1": Object {
              "label": "A",
            },
          }
        }
      />
    }
    onVisibleChange={[Function]}
    overlayClassName="TraceDiffHeader--popover"
    placement="bottomLeft"
    title={
      <TraceIdInput
        selectTrace={[Function]}
      />
    }
    trigger="click"
    visible={false}
  >
    <div
      className="ub-flex u-flex-1"
    >
      <TraceHeader />
    </div>
  </Popover>
</header>
`;

exports[`TraceDiffHeader renders as expected 1`] = `
<header
  className="TraecDiffHeader"
>
  <div
    className="TraecDiffHeader--labelItem-darkened"
  >
    <h1
      className="ub-m0"
    >
      A
    </h1>
  </div>
  <Popover
    content={
      <CohortTable
        cohort={
          Array [
            Object {
              "data": Object {
                "duration": 0,
                "startTime": 0,
                "traceName": "cohort-trace-name-0",
              },
              "error": "error 0",
              "id": "cohort-id-0",
              "state": "FETCH_ERROR",
            },
            Object {
              "data": Object {
                "duration": 100,
                "spans": Array [
                  Object {
                    "spanID": "trace-1-span-0",
                  },
                ],
                "startTime": 100,
                "traceName": "cohort-trace-name-1",
              },
              "error": "error 1",
              "id": "cohort-id-1",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 200,
                "spans": Array [
                  Object {
                    "spanID": "trace-2-span-1",
                  },
                  Object {
                    "spanID": "trace-2-span-2",
                  },
                ],
                "startTime": 200,
                "traceName": "cohort-trace-name-2",
              },
              "error": "error 2",
              "id": "cohort-id-2",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 300,
                "spans": Array [
                  Object {
                    "spanID": "trace-3-span-1",
                  },
                  Object {
                    "spanID": "trace-3-span-2",
                  },
                  Object {
                    "spanID": "trace-3-span-3",
                  },
                ],
                "startTime": 300,
                "traceName": "cohort-trace-name-3",
              },
              "error": "error 3",
              "id": "cohort-id-3",
              "state": "FETCH_DONE",
            },
          ]
        }
        current="cohort-id-1"
        selectTrace={[Function]}
        selection={
          Object {
            "cohort-id-1": Object {
              "label": "A",
            },
            "cohort-id-2": Object {
              "label": "B",
            },
          }
        }
      />
    }
    onVisibleChange={[Function]}
    overlayClassName="TraceDiffHeader--popover"
    placement="bottomLeft"
    title={
      <TraceIdInput
        selectTrace={[Function]}
      />
    }
    trigger="click"
    visible={false}
  >
    <div
      className="ub-flex u-flex-1"
    >
      <TraceHeader
        duration={100}
        error="error 1"
        startTime={100}
        state="FETCH_DONE"
        totalSpans={1}
        traceID="cohort-id-1"
        traceName="cohort-trace-name-1"
      />
    </div>
  </Popover>
  <div
    className="TraecDiffHeader--labelItem"
  >
    <h2
      className="u-tx-muted ub-mx3 ub-my0"
    >
      VS
    </h2>
  </div>
  <div
    className="TraecDiffHeader--labelItem-darkened"
  >
    <h1
      className="ub-m0"
    >
      B
    </h1>
  </div>
  <Popover
    content={
      <CohortTable
        cohort={
          Array [
            Object {
              "data": Object {
                "duration": 0,
                "startTime": 0,
                "traceName": "cohort-trace-name-0",
              },
              "error": "error 0",
              "id": "cohort-id-0",
              "state": "FETCH_ERROR",
            },
            Object {
              "data": Object {
                "duration": 100,
                "spans": Array [
                  Object {
                    "spanID": "trace-1-span-0",
                  },
                ],
                "startTime": 100,
                "traceName": "cohort-trace-name-1",
              },
              "error": "error 1",
              "id": "cohort-id-1",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 200,
                "spans": Array [
                  Object {
                    "spanID": "trace-2-span-1",
                  },
                  Object {
                    "spanID": "trace-2-span-2",
                  },
                ],
                "startTime": 200,
                "traceName": "cohort-trace-name-2",
              },
              "error": "error 2",
              "id": "cohort-id-2",
              "state": "FETCH_DONE",
            },
            Object {
              "data": Object {
                "duration": 300,
                "spans": Array [
                  Object {
                    "spanID": "trace-3-span-1",
                  },
                  Object {
                    "spanID": "trace-3-span-2",
                  },
                  Object {
                    "spanID": "trace-3-span-3",
                  },
                ],
                "startTime": 300,
                "traceName": "cohort-trace-name-3",
              },
              "error": "error 3",
              "id": "cohort-id-3",
              "state": "FETCH_DONE",
            },
          ]
        }
        current="cohort-id-2"
        selectTrace={[Function]}
        selection={
          Object {
            "cohort-id-1": Object {
              "label": "A",
            },
            "cohort-id-2": Object {
              "label": "B",
            },
          }
        }
      />
    }
    onVisibleChange={[Function]}
    overlayClassName="TraceDiffHeader--popover"
    placement="bottomLeft"
    title={
      <TraceIdInput
        selectTrace={[Function]}
      />
    }
    trigger="click"
    visible={false}
  >
    <div
      className="ub-flex u-flex-1"
    >
      <TraceHeader
        duration={200}
        error="error 2"
        startTime={200}
        state="FETCH_DONE"
        totalSpans={2}
        traceID="cohort-id-2"
        traceName="cohort-trace-name-2"
      />
    </div>
  </Popover>
</header>
`;
